File: src/PushRelabel.cpp
#include "PushRelabel.h"
#include <cassert>
#include "util.h"
#include <limits>
#include <iostream>

PushRelabel::Edge::Edge(int u, int v, double c) :
	origin(u),
	destination(v),
	capacity(c),
	flow(0)
{
}

template<class Predicate>
PushRelabel::eset PushRelabel::forPredicate(Predicate P) const {
	eset result;
	for (eptr edge : edges) {
		if (P(edge)) {
			result.insert(edge);
		}
	}
	return result;
}

PushRelabel::iset PushRelabel::getActive(const int source) const {
	iset activeNodes;
	for (int i : nodes()) {
		// The source is never an active node,
		// and also has a negative flow.
		if (i == source) continue;
		double flow = getExcess(i);

		// Sanity check
		assert(flow >= 0 && "Flow can never be negative!");
		
		if (flow > 0) {
			activeNodes.insert(i);
		}
	}
	return activeNodes;
}

double PushRelabel::getExcess(int i) const {
	// Since there is negative incoming flow for back edges, counting
	// the incoming flow is sufficient.
	eset incoming = getIncoming(i);
	return accumulate(incoming.begin(), incoming.end(), 0.0, [](double flow, const eptr& e) {
			return flow + e->flow;
	});
}

PushRelabel::eset PushRelabel::getOutgoing(int u) const {
	return forPredicate([u](eptr e) -> bool { return e->origin == u; });
}

PushRelabel::eset PushRelabel::getIncoming(int v) const {
	return forPredicate([v](const eptr& e) -> bool { return e->destination == v; });
}

void PushRelabel::addEdge(int u, int v, double c) {
	edges.insert(make_shared<Edge>(u, v, c));
	edges.insert(make_shared<Edge>(v, u, 0));
}

void PushRelabel::clear() {
	edges.clear();
}

void PushRelabel::init() {
	for (eptr edge : edges) {
		edge->flow = 0;
	}
	heights.clear();
}

double PushRelabel::maxFlow(const int s, const int t) {
	init();
	// Saturate outgoing edges of source
	for (eptr edge : getOutgoing(s)) {
		push(edge, numeric_limits<double>::infinity());
	}
	heights[s] = numNodes();

	iset active = getActive(s);
	while (!active.empty()) {
		const int node = randomElement(active);
		if (node == t) {
			active.erase(node);
			continue;
		}
		double excess = getExcess(node);
		assert(excess > 0);
		bool pushed = false;
		for (eptr edge : getOutgoing(node)) {
			if (canPush(edge) && excess) {
				excess -= push(edge, excess);
				pushed = true;

				if (excess == 0) {
					break;
				}
			}
		}

		if (!pushed) {
			// If we did not push, we should attempt a relabel.
			relabel(node);
		}

		active = getActive(s);
	}
	
	return getFlow(t);
}

PushRelabel::eptr PushRelabel::getBackEdge(const eptr& edge) const {
	eset candidates = forPredicate([edge](const eptr other) -> bool { return edge->origin == other->destination && edge->destination == other->origin; });
	assert(candidates.size() >= 1);

	return *candidates.begin();
}

bool PushRelabel::canPush(eptr edge) {
	int originHeight = heights[edge->origin];
	int destinationHeight = heights[edge->destination];

	return originHeight == destinationHeight + 1 && edge->capacity > edge->flow;
}

double PushRelabel::push(eptr edge, double amount) {
	double delta = min(amount, edge->capacity - edge->flow);
	eptr inverse = getBackEdge(edge);
	edge->flow += delta;
	inverse->flow -= delta;
	excess[edge->origin] -= delta;
	excess[edge->destination] += delta;

	return delta;
}

void PushRelabel::relabel(int node) {
	int newHeight = numeric_limits<int>::max();
	for (const eptr& edge : getOutgoing(node)) {
		if (edge->capacity > edge->flow) {
			newHeight = min(newHeight, heights[edge->destination]);
		}
	}

	heights[node] = newHeight + 1;
}

double PushRelabel::getFlow(const int t) const {
	double flow = 0;
	for (const eptr& edge : getIncoming(t)) {
		flow += edge->flow;
	}

	return flow;
}

int PushRelabel::numNodes() const {
	return (int) nodes().size();
}

set<int> PushRelabel::nodes() const {
	set<int> nodes;
	for (eptr e : edges) {
		nodes.insert(e->origin);
		nodes.insert(e->destination);
	}
	return nodes;
}

File: src/python/pushrelabel.py
from graph import Edge, Algorithm
import random

class PushRelabel(Algorithm):

    def __init__(self):
        Algorithm.__init__(self)

    def init(self, source, sink):
        self.D = dict()
        self.H = dict()

        # Initialze Height and Excess
        for node in self.V:
            self.H[node] = 0
            self.D[node] = 0

        self.H[source] = len(self.V)

        # Ensure reflectivity of edges
        for edge in self.E:
            self.F[edge] = 0
            backEdge = self.getBackEdge(edge)
            
            # Create the backedge if it does not exist.
            if not backEdge in self.C:
                self.C[backEdge] = 0
                self.F[backEdge] = 0
                self.E.append(backEdge)

        # Perform the saturating initial push
        for edge in self.E:
            if edge.u == source:
                self.push(edge, float("inf")) # Should be a better solution

    def push(self, edge, amount):
        delta = max(0, min(amount, self.C[edge] - self.F[edge]))
        backEdge = self.getBackEdge(edge)
        self.F[edge] += delta
        self.F[backEdge] -= delta
        self.D[edge.u] -= delta
        self.D[edge.v] += delta

    def relabel(self, node):
        minHeight = float("inf")
        for edge in self.E:
            if edge.u == node and self.C[edge] > self.F[edge]:
                minHeight = min(minHeight, self.H[edge.v] + 1)

        self.H[node] = minHeight

    def done(self):
        for node in self.D:
            if self.D[node] > 0 and node != self.sink:
                return False

        return True

    def iter(self):
        node = self.getActiveNode()
        edge = self.getPushableEdge(node)

        if edge == None:
            print "Relabeling " + str(node)
            self.relabel(node)
        else:
            print "Pushing edge " + str(edge)
            self.push(edge, self.D[node])

    def getActiveNode(self):
        nodes = list(self.V)
        random.shuffle(nodes)

        for node in nodes:
            if self.D[node] > 0 and node != self.sink:
                return node

        assert False

    def getPushableEdge(self, node):
        edges = list(self.E)
        random.shuffle(edges)
        for edge in edges:
            if edge.u == node and self.H[edge.u] == self.H[edge.v] + 1 and self.C[edge] > self.F[edge]:
                return edge

        return None

    def print_status(self):
        Algorithm.print_status(self)
        for node in self.V:
            print "Node " + str(node) + " has height " + str(self.H[node])


if __name__ == '__main__':
    G = PushRelabel()
    G.test()

    print "Completed."
